### PR TITLE
Melee and Box pickup detection fixed

### PR DESCRIPTION
**Pull request by mikegreber**

### CODE REVIEW

File: HITDAD/js/plugins/SPFBoxMouseToss.js [hunk 1]
```diff
@@ -121,15 +111,58 @@
...
+        events.forEach(function(event) {
+            if (event._canPickup) {
+                pickupableEvents.push(event);
+            }
+        });
+
+        return pickupableEvents;
+    }
+
+    // Checks if and returns an enemy that is within melee range, returns null if none in range.
```
Comment (vppatel111) [inline HITDAD/js/plugins/SPFBoxMouseToss.js]: Comment is talking about enemies but this function is about boxes?

File: HITDAD/js/plugins/SPFBoxMouseToss.js [hunk 2]
```diff
@@ -121,15 +111,58 @@
...
+    function findBoxWithinReach() {
+        let direction = $gamePlayer.direction();
+
+        // So it scans from around the players back for detecting objects underneath
+        let xTraceStart = direction === 4 ? $gamePlayer._realX + 0.5 : $gamePlayer._realX  - 0.5;
```
Comment (vppatel111) [inline HITDAD/js/plugins/SPFBoxMouseToss.js]: `let xTraceStart = direction === DIRECTION.LEFT ? $gamePlayer._realX + 0.5 : $gamePlayer._realX  - 0.5;`

=> Which is defined in SPFCore.js

File: HITDAD/js/plugins/SPFBoxMouseToss.js [hunk 3]
```diff
@@ -121,15 +111,58 @@
...
+
+        let closestBox = null;
+        let closestBoxDistance = null;
+
+        SPF_Boxes.forEach(function(box) {
+            let distanceToBox =  box._realX - xTraceStart; // Will be positive if box is to right of player
+            let verticalOffset = Math.abs($gamePlayer._realY - box._realY);
+
+            // Get box in direction player is facing to have positive distance value (direction === 6 means right)
+            let forwardDistanceToBox = direction === 6 ? distanceToBox : - distanceToBox;
```
Comment (vppatel111) [inline HITDAD/js/plugins/SPFBoxMouseToss.js]: In this case, you could simply use Math.abs(distanceToBox) to get the absolute value of distance to box.

File: HITDAD/js/plugins/SPFMeleeAttack.js [hunk 4]
```diff
@@ -38,102 +38,131 @@
...
+  //
+  //     if (event._npcType === SPF_NPCS.NORMAL_GUARD) {
+  //       enemyEvent = event;
+  //     }
+  //   });
+  //
+  //   return enemyEvent;
+  // }
+
+  // Checks if and returns an enemy that is within melee range.
```
Comment (vppatel111) [inline HITDAD/js/plugins/SPFMeleeAttack.js]: Don't worry about commenting out code from now on, just delete it if your replacing it.

File: HITDAD/js/plugins/SPFMeleeAttack.js [hunk 5]
```diff
@@ -38,102 +38,131 @@
...
+    let closestEnemy = null;
+    let closestEnemyDistance = null;
+
+    SPF_Enemies.forEach(function(enemy) {
+      if (!enemy._npcType || SPF_IsEnemyPacified(enemy)) return;
+      let distanceToEnemy = $gamePlayer._realX - enemy._realX; // Will be negative if enemy is on right of player
+      let verticalOffset = Math.abs($gamePlayer._realY - enemy._realY);
+
+      // Get enemies in direction player is facing to have positive distance value (direction === 6 means right)
+      let forwardDistanceToEnemy = direction === 4 ?  distanceToEnemy : -distanceToEnemy;
```
Comment (vppatel111) [inline HITDAD/js/plugins/SPFMeleeAttack.js]: Similarly Math.abs() here.

File: HITDAD/js/plugins/SPFMeleeAttack.js [hunk 6]
```diff
@@ -38,102 +38,131 @@
...
+    //     }
+    //     break;
+    //   case 6: // Right
+    //     if ($gamePlayer.isCollidedWithEvents($gamePlayer.x + 1, $gamePlayer.y)) {
+    //       return $gameMap.eventsXyNt($gamePlayer.x + 1, $gamePlayer.y);
+    //     }
+    //     break;
+    // }
+
+    // return []; // No collision with events.
```
Comment (vppatel111) [inline HITDAD/js/plugins/SPFMeleeAttack.js]: Remove commented code entirely.

File: HITDAD/js/plugins/SPFMeleeAttack.js [hunk 7]
```diff
@@ -38,102 +38,131 @@
...
-      if (!SPF_isEmpty(enemyHit)) {
-        $gameParty.loseItem(item, 1);
-        SPF_IncapacitateEnemy(enemyHit);
-        AudioManager.playSe(sePacify);
-      }
 
-    }
 
-  }
+  // function PerformAttack() {
```
Comment (vppatel111) [inline HITDAD/js/plugins/SPFMeleeAttack.js]: You get the gist be now.

File: HITDAD/js/plugins/SPFBoxMouseToss.js (hunk 3 above)
Comment (mikegreber) [reply HITDAD/js/plugins/SPFBoxMouseToss.js]: Math.abs does not work in this case as negative values are intended for objects behind the player, this function has been refactored to be easier to understand.

**Comment by vppatel111** [comment]
Tested and looks amazing!

**Pull request merged**